Istio 1.18 support (#8676)
solo-io/gloo: docs/content/guides/integrations/service_mesh/istio.md

Complete the following tasks before configuring an Istio sidecar for your Gloo Edge gateway: 

> 1. Create or use an existing cluster that runs Kubernetes version 1.24 or later. 

Author: I'm not sure when/why we drop support for k8s versions, and in fact don't remember exactly I mentioned k8s 1.24 as the minimum, but likely because that's the minimum supported by Istio 1.18
I can clarify that it needs to be a version compatible with both the Gloo version and Istio version

> 1. Create or use an existing cluster that runs Kubernetes version 1.24 or later. 

Reviewer: > I can clarify that it needs to be a version compatible with both the Gloo version and Istio version
Saw the changes, looks good! 👍🏼

   helm repo update
   ```
      
3. Create a `value-overrides.yaml` file with the following content:
- To configure your gateway with an Istio sidecar, add the `istioIntegration` section and set `enableIstioSidecarOnGateway` option to `true`
- Set `disableAutoInjection` to `true` in order to ensure Gloo components are not included in the Mesh
- Set `global.istioSDS.enabled` to `true` to allow the Gateway Proxy to consume Istio certs despite not being in the Mesh 
- Specify image fields under `global.glooMtls.istioProxy.image` and `global.glooMtls.sds.image` corresponding with the version of Istio and Gloo Edge installed respectively
  - The default Istio version is 1.18.2
   ```yaml
   global:
     istioIntegration:
       enableIstioSidecarOnGateway: true

>        disableAutoInjection: true

Author: I allude to it above [on line 34](https://github.com/solo-io/gloo/pull/8676/files#diff-dd73333b0d09e443e1df8c587599327a58bfb10f2e672e52c0cf819140af112bR34)

Other than that I don't think there is documentation

I could add a sub-bullet below 34 like
```
  - Requiring mTLS for traffic between Gloo components in a mesh can break functionality
```

Beyond that may be beyond the scope for this PR but I'm open to suggestions

>        disableAutoInjection: true

Reviewer 2: Yeah I was just thinking of a comment on the definition in values.go where the Helm API is defined, so try to call it out in code

>        disableAutoInjection: true

Author: The current description for that value is as follows:

> Annotate all pods (excluding those whitelisted by other config values) to with an explicit 'do not inject' annotation to prevent Istio from adding sidecars to all pods. It's recommended that this be set to true if Gloo's namespace is marked for Istio discovery, as some pods do not immediately work with an Istio sidecar without extra manual configuration.

Perhaps its worth removing "if Gloo's namespace is marked for Istio discovery" and leaving it as more of a blanket recommendation?

>        disableAutoInjection: true

Reviewer 2: Yeah I like that

   ```
      

> 3. Create a `value-overrides.yaml` file with the following content. To configure your gateway with an Istio sidecar, make sure to add the `istioIntegration` section and set the `enableIstioSidecarOnGateway` option to `true`. You can optionally add the `global.istioSDS.enabled` option to your overrides file to automatically renew the certificate that the sidecar uses before it expires. 
> Be sure to specify valid image fields under `global.glooMtls.istioProxy.image` and `global.glooMtls.sds.image`. The default Istio version is 1.18.2.

Reviewer 3: What does the valid image statement mean? Does it just mean they should always use these images we recommend, or is here something else they need to verify for the istioproxy and sds images? We might be able to do something like this:
```suggestion
3. Create a `value-overrides.yaml` file with the following content.
   * To configure your gateway with an Istio sidecar, make sure to add the `istioIntegration` section and set the `enableIstioSidecarOnGateway` option to `true`.
   * You can optionally add the `global.istioSDS.enabled` option to your overrides file to automatically renew the certificate that the sidecar uses before it expires. 
   * The default values are used for the `global.glooMtls.istioProxy.image` and `global.glooMtls.sds.image` options. You can change these default values to use other valid images as needed.
```

Author: I want to keep the suggested values that are copy/pastable, but happy to use bullet points to call out what's important here if that's cleaner

To clarify as to what's "valid":
- For the Istio image it should match the version of Istio that's installed on the cluster
- For the SDS image it should match the version of Gloo that's installed

Also `istioSDS.enabled` is required for this guide since it exposes the certs for the `gateway-proxy` to consume
Upstreams get configured by `glooctl istio enable-mtls --upstream <upstream name>` to point to the gateway proxy's SDS container
(I realize this is leftover from the previous guide and I neglected to remove it)

   ```yaml
   global:
     istioIntegration:
       enableIstioSidecarOnGateway: true
       disableAutoInjection: true
     istioSDS:
       enabled: true
     glooMtls:
       istioProxy:
         image:
           registry: docker.io/istio
           repository: proxyv2
           tag: 1.18.2
       sds:
         image:
           repository: sds
           tag: 1.15.7
   gatewayProxies:
     gatewayProxy:
       podTemplate: 
         httpPort: 8080
         httpsPort: 8443
   ```

   {{% notice note %}}

>    Though we're actually configuring istioMtls in this guide, the `istioProxy` values for this feature are taken from the `glooMtls` block.

Reviewer 3: ```suggestion
   Although the `istioProxy` values are defined in the `glooMtls` block, the values are used to configure `istioMtls`.
```

   {{% /notice %}}
   
4. Install or upgrade Gloo Edge. 
   {{< tabs >}} 
   {{< tab name="Install Gloo Edge">}}

   Install Gloo Edge with the settings in the `value-overrides.yaml` file. This command creates the `gloo-system` namespace and installs the Gloo Edge components into it.
   ```shell
    helm install gloo gloo/gloo --namespace gloo-system --create-namespace -f value-overrides.yaml
   ```
   {{< /tab >}}
   {{< tab name="Upgrade Gloo Edge">}}
   Upgrade Gloo Edge with the settings in the `value-overrides.yaml` file.
   ```shell
   helm upgrade gloo gloo/gloo --namespace gloo-system -f value-overrides.yaml
   ```
   {{< /tab >}}
   {{< /tabs >}}   
5. [Verify your setup]({{< versioned_link_path fromRoot="/installation/gateway/kubernetes/#verify-your-installation" >}}). 
   

> 6. Get the pods for your gateway proxy deployment. You should now see three containers in the gateway-proxy pod. 

Reviewer 3: ```suggestion
6. Verify that your `gateway-proxy` pod now has three containers. 
```

   ```shell
   kubectl get pods -n gloo-system
   ```
    
   Example output: 
   ```
   NAME                             READY   STATUS      RESTARTS   AGE
   discovery-6dcc8ddc58-q4zv7       1/1     Running     0          39s
   gateway-certgen-xzr7t            0/1     Completed   0          43s
   gateway-proxy-7bc5c97449-n9498   3/3     Running     0          39s
   gloo-d8cfbf86b-v59j4             1/1     Running     0          39s
   gloo-resource-rollout-hhvf9      0/1     Completed   0          38s
   ```
    

> 7Describe the `gateway-proxy` pod to verify that the `istio-proxy` and `sds` containers are running. 

Reviewer 3: ```suggestion
7. Describe the `gateway-proxy` pod to verify that the `istio-proxy` and `sds` containers are running.
```

   ```shell
   kubectl describe <gateway-pod-name> -n gloo-system
   ```

Congratulations! You have successfully configured an Istio sidecar for your Gloo Edge gateway. 

## Set up and verify the mTLS connection 

> To demonstrate that you can connect to your app via mutual TLS (mTLS), we will: 

Reviewer 3: ```suggestion
To demonstrate that you can connect to your app via mutual TLS (mTLS), you can follow these steps to:
```

- Install the Bookinfo app in your cluster

> - Set up a virtual service to route incoming requests to the

Reviewer 3: im guessing for bookinfo
```suggestion
- Set up a virtual service to route incoming requests to the Bookinfo app
```

       enableIstioSidecarOnGateway: true
     istioSDS:
       enabled: true
     glooMtls:
       istioProxy:
         image:
           registry: docker.io/istio
           repository: proxyv2
           tag: 1.18.2
       sds:
         image:
           repository: sds

>            tag: 1.15.7

Author: I describe what these versions should be [on line 36](https://github.com/solo-io/gloo/pull/8676/files#diff-dd73333b0d09e443e1df8c587599327a58bfb10f2e672e52c0cf819140af112bR36)

If that's not clear or easy to miss I'm happy to add inline comments in this YAML block

>            tag: 1.15.7

Reviewer 2: Yeah I was thinking inline comment in the YAML, but I think what you have is enough too. This LGTM

>            tag: 1.0.0-ci

Author: If `istioSDS` is enabled, as it is already in the guide, image fields are derived from the `glooMtls` `istioProxy` and `sds` sections

I'm not actually sure what SDS values should be, as `1.0.0-ci` is not likely available to users, but this is probably worth addressing and updating before merge

I'm not actually sure why the `istioProxy` image values are relevant as Istio overwrites the container we write if Istio injection is enabled for the pod

>            tag: 1.0.0-ci

Author: To close the loop here, SDS is released with every Gloo OSS release and therefore the version used should match the minor version of Gloo being used, if not the exact version

   serviceaccount/bookinfo-productpage created
   deployment.apps/productpage-v1 created
   ```

2. Create a virtual service to set up the routing rules for your Bookinfo app. In the following example, you instruct the Gloo Edge gateway to route incoming requests on the `/productpage` path to be routed to the `productpage` service in your cluster. 
   ```yaml
   kubectl apply -f- <<EOF
   apiVersion: gateway.solo.io/v1
   kind: VirtualService
   metadata:
     name: vs
     namespace: gloo-system
   spec:
     virtualHost:
       domains:
       - 'www.example.com'
     routes:
     - matchers:
       - prefix: /productpage
       routeAction:
         single:
           upstream:
             name: default-productpage-9080
             namespace: gloo-system
   EOF
   ```
   

> 3. At this point we can send a request to the product page. 

Reviewer 3: ```suggestion
3. Test access to Bookinfo by sending a request to the product page. 
```

   ```shell
   curl -viks -H "Host: www.example.com" "$(glooctl proxy url)/productpage" --output /dev/null
   ```

>    ```text

Reviewer 3: ```suggestion
   A `200 OK` response indicates that Bookinfo is reachable. However, traffic to the product page upstream is not encrypted with mTLS.
   ```text
```

   *   Trying 127.0.0.1:32000...
   * Connected to localhost (127.0.0.1) port 32000 (#0)
   > GET /productpage HTTP/1.1
   > Host: www.example.com
   > User-Agent: curl/7.88.1
   > Accept: */*
   >
   < HTTP/1.1 200 OK
   < content-type: text/html; charset=utf-8
   < content-length: 5290
   < server: envoy
   < date: Tue, 03 Oct 2023 20:14:00 GMT
   < x-envoy-upstream-service-time: 17
   <
   { [5290 bytes data]
   * Connection #0 to host localhost left intact
   ```

>    A 200 response indicates that we can reach our app, however if we stop here, traffic to the productpage Upstream will not be encrypted with mTLS.

Reviewer 3: moved up above the output
```suggestion
```

> 4. In order to require all traffic in the Mesh uses mTLS, apply the following STRICT PeerAuthentication policy:

Reviewer 3: ```suggestion
4. To require all traffic in the mesh to use mTLS, apply the following STRICT PeerAuthentication policy.
```

   ```yaml
   kubectl apply -f - <<EOF
   apiVersion: "security.istio.io/v1beta1"
   kind: "PeerAuthentication"
   metadata:
     name: "test"
     namespace: "istio-system"
   spec:
     mtls:
       mode: STRICT
   EOF
   ```

> 5. Now if we make the same curl request we will get a 503 response, as our upstream is not configured for Istio mTLS.

Reviewer 3: ```suggestion
5. Repeat the same curl request to the product page.
```

   ```shell
   curl -viks -H "Host: www.example.com" "$(glooctl proxy url)/productpage" --output /dev/null 
   ```

>    ```text

Reviewer 3: ```suggestion
   The `503 Service Unavailable` response is returned because the upstream is not configured for Istio mTLS.
   ```text
```

   *   Trying 127.0.0.1:32000...
   * Connected to localhost (127.0.0.1) port 32000 (#0)
   > GET /productpage HTTP/1.1
   > Host: www.example.com
   > User-Agent: curl/7.88.1
   > Accept: */*
   >
   < HTTP/1.1 503 Service Unavailable
   < content-length: 95
   < content-type: text/plain
   < date: Tue, 03 Oct 2023 20:15:55 GMT
   < server: envoy
   <
   { [95 bytes data]
   * Connection #0 to host localhost left intact
   ```

> 6. Use `glooctl` to configure the upstream for Istio mTLS: 

Reviewer 3: ```suggestion
6. Use `glooctl` to configure the upstream for Istio mTLS.
```

   ```shell
   glooctl istio enable-mtls --upstream default-productpage-9080
   ```

> 7. Now the request will once again succeed, with a 200 response, with traffic encrypted using mTLS.

Reviewer 3: ```suggestion
7. Repeat the same curl request to the product page. This time, the request succeeds because traffic is encrypted using mTLS.
```
